state-sync/src/coordinator.rs

```diff
@@ -811,92 +869,97 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
         Ok(target_li)
     }
 
-    /// Applies (= executes and stores) chunk to storage if `response` is valid
-    /// Chunk response checks performed:
-    /// - does chunk contain no transactions?
-    /// - does chunk of transactions matches the local state's version?
-    /// - verify LIs in chunk response against local state
-    /// - execute and commit chunk
-    /// Returns error if above chunk response checks fail or chunk was not able to be stored to storage, else
-    /// return Ok(()) if above checks all pass and chunk was stored to storage
-    fn apply_chunk(&mut self, peer: &PeerNetworkId, response: GetChunkResponse) -> Result<()> {
+    /// Applies (i.e., executes and stores) the chunk to storage iff `response` is valid.
+    fn apply_chunk(
+        &mut self,
+        peer: &PeerNetworkId,
+        response: GetChunkResponse,
+    ) -> Result<(), Error> {
         debug!(
             LogSchema::event_log(LogEntry::ProcessChunkResponse, LogEvent::Received)
                 .chunk_response(response.clone())
                 .peer(peer)
         );
         fail_point!("state_sync::apply_chunk", |_| {
-            Err(anyhow::anyhow!("Injected error in apply_chunk"))
+            Err(crate::error::Error::UnexpectedError(
+                "Injected error in apply_chunk".into(),
+            ))
         });
 
+        // Check response comes from upstream peer
         if !self.request_manager.is_known_upstream_peer(peer) {
             counters::RESPONSE_FROM_DOWNSTREAM_COUNT
                 .with_label_values(&[
                     &peer.raw_network_id().to_string(),
                     &peer.peer_id().to_string(),
                 ])
                 .inc();
-            bail!("received chunk response from downstream");
+            return Err(Error::ReceivedChunkFromDownstream(peer.to_string()));
         }
 
+        // Check chunk is not empty
         let txn_list_with_proof = response.txn_list_with_proof.clone();
-        let known_version = self.local_state.synced_version();
         let chunk_start_version =
             txn_list_with_proof
                 .first_transaction_version
                 .ok_or_else(|| {
                     self.request_manager.process_empty_chunk(&peer);
-                    format_err!("[state sync] Empty chunk from {:?}", peer)
+                    Error::ReceivedEmptyChunk(peer.to_string())
                 })?;
 
+        // Check chunk starts at the correct version
+        let known_version = self.local_state.synced_version();
         let expected_version = known_version
             .checked_add(1)
-            .ok_or_else(|| format_err!("Expected version has overflown!"))?;
+            .ok_or_else(|| Error::IntegerOverflow("Expected version has overflown!".into()))?;
```

Comment:
Interesting, why not take an &str, if the integer overflow doesn't take any variable string?

Reply:
Mostly for consistency. Whenever we've done this custom error::Error type, we tend to use Strings as opposed to &str. I don't think there's a fundamental reason why. It often tends to be shorter in the code (as opposed to sometimes having to call weird methods (e.g., `as_str()`, or `&object.to_string()`)

Reply:
Yeah, I think we mostly use `String` so we can inject in extra information like identifiers and stuff